azure: Fix allocation of addresses

tgraf opened the pull request:
Fixes: 9779b09b6a ("ipam: Support for Azure IPAM")

ungureanuvladvictor commented on pkg/azure/api/api.go:
```diff
@@ -61,6 +65,7 @@ func NewClient(subscriptionID, resourceGroup string, metrics MetricsAPI, rateLim
 		resourceGroup:   resourceGroup,
 		interfaces:      network.NewInterfacesClient(subscriptionID),
 		virtualnetworks: network.NewVirtualNetworksClient(subscriptionID),
+		vms:             compute.NewVirtualMachineScaleSetVMsClient(subscriptionID),
```
mind naming this `vmssVMs` since we'll have to add support for plain VMs later as well?

ungureanuvladvictor commented on pkg/azure/types/types.go:
```diff
@@ -85,13 +92,64 @@ type AzureInterface struct {
+	vmssName string
+
+	// vmName is the name of the virtual machine
+	vmName string
```
mind renaming this to `vmID` ? I realize the setup is going to be a bit weird when we add support for plain VMs because in that case `vmssName` + `vmID` will be empty but we'll have a valid `vmName` and would like to deconflict now.

tgraf replied:
But isn't this the name so we can reuse for VMs? I assume the vmID will be fully qualified.

```
subscriptions/xxx/resourceGroups/yyy/providers/Microsoft.Compute/virtualMachineScaleSets/ssss/virtualMachines/vvv/networkInterfaces/iii
```

`vmName` is currently et to `vvv`

ungureanuvladvictor replied:
I think the nomenclature in azure is a bit different:
- for plain VMs the vmName is what you see in the console at the machine
- for VMSS the vmName will be `$vmssName-vvvvvv` (https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-instance-ids)

In this case for VMSS we treat the vmID to be vmName which does not align with how Azure wants to think about things.

Don't want to over index on this, we can always re-work this code a bit later.

tgraf replied:
OK. That makes sense. I've renamed it.

ungureanuvladvictor reviewed (commented):
just some naming things for now.

ungureanuvladvictor commented on pkg/azure/api/api.go:
```diff
@@ -274,32 +288,50 @@ func (c *Client) GetVpcsAndSubnets(ctx context.Context) (ipamTypes.VirtualNetwor
+	netIfConfig.IPConfigurations = &ipConfigurations
 
-	_, err := c.interfaces.CreateOrUpdate(ctx, c.resourceGroup, interfaceID, ifaceParams)
+	_, err = c.vmss.Update(ctx, c.resourceGroup, vmssName, instanceID, result)
```
I think we need to wait for this `Update` to complete on the Azure backend otherwise the allocator will kick again at a re-sync and it will double allocate.

`Update` returns a `VirtualMachineScaleSetVMsUpdateFuture` on which you can call https://github.com/Azure/go-autorest/blob/master/autorest/azure/async.go#L144-L152 and that will block until it finishes.

In my testing I was not able to repro the double alloc after I deployed a modified version of your PR.

ungureanuvladvictor commented on pkg/azure/api/l:
```diff
@@ -0,0 +1,69 @@
+level=info msg="Skipped reading configuration file" reason="Config File \"cilium-operator\" Not Found in \"[/]\"" subsys=config
```
Wrongly added I assume.

tgraf replied:
Yes, removed

ungureanuvladvictor commented on pkg/azure/api/node2:
```diff
@@ -0,0 +1,397 @@
+apiVersion: cilium.io/v2
```
Wrongly added I assume?

tgraf replied:
Yes, removed.

tgraf committed:
ipam: Add FirstSubnetWithAvailableAddresses

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
azure: Calculate available addresses based on subnet resource

Using the subnet resource allows to avoid using an allocator to scan all
used addresses.

Signed-off-by: Thomas Graf <thomas@cilium.io>

ungureanuvladvictor committed:
azure: Wait for AssignPrivateIpAddresses() to complete

This is an async operation, not waiting for it to complete will cause
recalculation of required addresses to happen too early and too many
addresses will be allocated.

Signed-off-by: Vlad Ungureanu <vladu@palantir.com>

tgraf committed:
cidr: Add AvailableIPs() function

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
azure: Fix address allocation

The CreateOrUpdate API did not support adding additional addresses. IPs
must be requested via the VMScaleSet.

Fixes: 9779b09b6a ("ipam: Support for Azure IPAM")

Signed-off-by: Vlad Ungureanu <vladu@palantir.com>
Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
Move RandomRune() pkg/rand/

This functionality will be needed outside of tests in the next commit so
move it out of testutils.

Signed-off-by: Thomas Graf <thomas@cilium.io>
